### DISCUSSION TITLE
Verifying Total Sojourn Time in a Specific State in UPPAAL

### DISCUSSION THREAD

**Discussion by Szpilman2** [UPPAALModelChecker/UPPAAL-Meta#271] [Q&A] [answered]
Dear UPPAAL Team,

I hope this message finds you well.

I have two timed automata named "Software" and "Hardware". I have attached pictures of the models. 
software:
![SoftwareTA](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/35931007/31366525-6770-4b97-9549-b25a53a861d4)
hardware:
![HardwareTA](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/35931007/9a9d06d3-83be-4633-89fd-4fceea256305)

Whenever the Software TA sends a signal through the "use" channel, the Hardware TA receives the signal and transitions to the "using" state. The Hardware remains in the "using" state (in my case 2 units of time) for a short period before transitioning to the "tail" state.

I would like to write a query that verifies whether the total time the Hardware spends in the "tail" state (accumulated over all instances of "use?" in the Software TA) is less than a certain constant. I am unsure how to write this query. My initial approach was to save the clock values, but I understand that this is not possible in UPPAAL.

In other words, I need to check if the sum of the sojourn times in the "tail" state is less than a specific constant.

I have also attached my model for your reference.
[EnergyHotspotModel.zip](https://github.com/UPPAALModelChecker/UPPAAL-Meta/files/15398506/EnergyHotspotModel.zip)

Thank you very much for your assistance.

Best regards,
-- Peyman

**Comment by mikucionisaau** [comment]
To extract the maximum time spent in `Tail`, try the supremum query:
```
sup{hardware.tail}: clk
```
To compute the cumulative duration, please add a stop watch, e.g. `sw`: a `clock` whose progress is stopped with  `sw'==0` expression in the invariant (add such invariant to all locations of `hardware` except `Tail`) and then use the supremum query over `sw`.

Note that with stop watches the symbolic analysis becomes over-approximate (UPPAAL may start issuing `maybe` verdicts to certain queries), but in such case it's most likely still exact (I don't have a proof) as it does not disturb the logic of the model (model does not use constraints over `sw`).

See the publication about stopwatches: https://doi.org/10.1007/3-540-44618-4_12

**Comment by Szpilman2** [comment]
thank you so much. your answers are life saving.
you mean to say I define ```hybrid clock sw``` in HardwareModel declarations and add this to all location of hardware except Tail like this?

![newModel](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/35931007/638a3192-d9ba-478d-95dc-75dde8624c9b)

 now my question is some locations like ```Using``` has invariant like ```clk <= D_Use && sw' == 0``` and I think it is not correct. in other words adding ```sw``` to the invariants affects my original invariant in the model. how can I figure this out?

also is ```sup{}:sw``` correct query for my property?

**Reply by Szpilman2** [reply]
I also recall attempting to define a hybrid clock and calculate its supremum, but UPPAAL returned an error indicating that this operation is not supported. Is my understanding correct?

**Reply by mikucionisaau** [reply]
No, not `hybrid`, just plain `clock`. `hybrid` is for dynamical cost which can be ignored in symbolic analysis.

Your invariants are correct: the derivative constraint `sw'==0` says that it's only the derivative of `sw` is constrained, but not the value of `sw` (and neither the values of other clocks). Basically all the original constraints are kept.

The over-approximation bit comes into play due to fact that stopwatch automata reachability is undecidable in general and it requires additional constraints (involving three clocks), but DBM structure cannot represent them, so we get fewer constraints == over-approximation.

As for the query, use either:
```
sup: sw
sup{hardware.Tail}: sw
```
The first one is short, fast and precise enough in this case. The second is a bit superfluous, but it shows the intention.

**Reply by Szpilman2** [reply]
thank you very very much.
I have modified my model but although i have defined ```sw``` but UPPAAL gives me this error: ```unknown identifier sw```
![error](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/35931007/e9f19655-bd8f-4187-a737-ca52724f9b9c)

this is my modified model:
[EnergyHotspotModel.zip](https://github.com/UPPAALModelChecker/UPPAAL-Meta/files/15399958/EnergyHotspotModel.zip)


what shall I do?

thanks again.

**Reply by mikucionisaau** [reply]
See publication about stopwatches: https://doi.org/10.1007/3-540-44618-4_12

**Reply by mikucionisaau** [reply]
Did you declare the `sw`?
You need to have a line in global declarations, like:
```c
clock sw;
```

**Reply by Szpilman2** [reply]
no no I did not define it in global declaration I had defined it in declarations for Hardware TA.
now it's my result. UPPAAL says that ```the property may be satisfied``` and returns ```16``` as result of the query.

![fixedresult](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/35931007/80caed4f-1ab1-4ebe-8933-af64b7ed61cb)

now how can I compare result of ```sup sw``` with specific constant in another query?

Marius you are my hero man. Thank you so much.

**Reply by Szpilman2** [reply]
also I will check the publications thanks for sharing.

**Reply by mikucionisaau** [reply, marked as answer]
If you define `sw` locally in `hardware` then you can refer to it by `hardware.sw` just like any variable.
